libraries/llvm/buffer.c

```diff
@@ -0,0 +1,141 @@
+#ifndef EFFEKT_BUFFER_C
+#define EFFEKT_BUFFER_C
+
+
+// TODO It may be performance-advantageous to implement this C file's semantics
+// in LLVM, since the linker cannot realistically be asked to satisfactorily
+// inline.
+
+
+// Eight bytes for the reference counter.
+// E.g. capacity could also be stored left of the data.
+#define BUFFER_METADATA_WIDTH (8)
+
+uint64_t c_buffer_length(const struct Pos buffer) {
+    return buffer.tag;
+}
+
+// NOTE: This assumes a homogenously used byte order.
+uint64_t *c_buffer_refcount(const struct Pos buffer) {
+    return (uint64_t *) (buffer.obj - 8);
+}
+
+uint8_t *c_buffer_bytes(const struct Pos buffer) {
+    return buffer.obj;
+}
+
+
+struct Pos c_buffer_construct(const uint64_t n, const uint8_t *data) {
+    uint8_t *buf = malloc(BUFFER_METADATA_WIDTH + n * sizeof *buf);
+    ASSERT_NON_NULL(buf)
+
+    // reference count (a reference count of zero means one sole owner)
+    for (uint64_t j = 0; j < BUFFER_METADATA_WIDTH; ++j)
+        buf[j] = 0;
+
+    // data
+    for (uint64_t j = 0; j < n; ++j)
+        buf[BUFFER_METADATA_WIDTH + j] = data[j];
+
+    return (struct Pos) {
+        .tag = n,
+        .obj = BUFFER_METADATA_WIDTH + buf,
+    };
+}
+
+void c_buffer_destruct(const struct Pos buffer) {
+    free(buffer.obj - BUFFER_METADATA_WIDTH);
+}
+
+
+void c_buffer_refcount_increment(const struct Pos buffer) {
+    if (DEBUG_REFCOUNT) {
+        fprintf(stderr, "c_buffer_refcount_increment((struct Pos) "
+            "{ .tag = %" PRIu64 ", .obj = %p }): from %" PRIu64 "\n",
+            buffer.tag, buffer.obj, *(c_buffer_refcount(buffer)));
+        fflush(stderr);
+    }
+
+    (*c_buffer_refcount(buffer))++;
+}
+
+void c_buffer_refcount_decrement(const struct Pos buffer) {
+    if (DEBUG_REFCOUNT) {
+        fprintf(stderr, "c_buffer_refcount_decrement((struct Pos) "
+            "{ .tag = %" PRIu64 ", .obj = %p }): from %" PRIu64 "\n",
+            buffer.tag, buffer.obj, *(c_buffer_refcount(buffer)));
+        fflush(stderr);
+    }
+
+    if (!(*c_buffer_refcount(buffer))--)
+        c_buffer_destruct(buffer);
+}
+
+
+struct Pos c_buffer_construct_zeroed(const uint64_t n) {
+    uint8_t *zeroes = calloc(n, sizeof *zeroes);
+    ASSERT_NON_NULL(zeroes)
+    const struct Pos buffer = c_buffer_construct(n, zeroes);
+    free(zeroes);
+    return buffer;
+}
+
+void c_buffer_truncate(struct Pos buffer, const uint64_t n) {
+    if (n > c_buffer_length(buffer))
+        return;
+    buffer.tag = n;
+}
+
+struct Pos c_buffer_copy(const struct Pos buffer) {
+    return c_buffer_construct(c_buffer_length(buffer), c_buffer_bytes(buffer));
+}
+
+
+// incurs a malloc: the returned pointer needs to be managed
+char *c_buffer_as_null_terminated_string(const struct Pos buffer) {
+    // Zero runes are represented as non-minimal utf8 in the null-terminated
```

Comment:
Are these conversions to- and from null-terminated strings currently used? I am a bit afraid of responsibility for freeing memory.

Reply:
See the `feature/llvm-io` branch: https://github.com/effekt-lang/effekt/blob/abb6448d68b7d0c7e09d775d3e2319f9c0e41922/libraries/llvm/io.c#L72

Reply:
They are not exposed through `libraries/llvm/forwared-declare.ll` and meant for internal use only. You need to shed your fear of manual memory management if you intend to build complex objects onto a C-heritaged memory model.

Reply:
But if you want to banish said lines, feel free to do so.